File: mathewDavidov/assignments/mergeIntervals/lc57/lc57.java

/**
 * Problem Statement:
 * Given a list of non-overlapping intervals sorted by their start time, 
 * insert a given interval at the correct position and 
 * merge all necessary intervals to produce a list that has only mutually exclusive intervals.
 * 
 * Example 1:
 * Input: intervals = [[1,3],[6,9]], newInterval = [2,5]
 * Output: [[1,5],[6,9]]
 * 
 * Example 2:
 * Input: intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]], newInterval = [4,8]
 * Output: [[1,2],[3,10],[12,16]]
 * Explanation: Because the new interval [4,8] overlaps with [3,5],[6,7],[8,10].
 * 
 * Edge cases:
 * null / empty array -> return new interval
 * 
 * The problem is similar to merged intervals (LC 56) with only one added interval.
 * Therefore, we can approach it in a similar fashion, but begin with the start and end
 * as the new interval's start and end. 
 * We must also modify the logic in the for loop to account for normal non-overlapping intervals.
 * 
 * initialize the start to the new interval's start
 * initialize the end to the new interval's end
 * for each interval:
 *   if the current interval's end is less than the start (i.e. unaffected by the new interval):
 *     create an interval containing the current start and end and add it to the merged list
 *   else if the current interval's start is less than or equal to the end:
 *     update the start to be the min of the current interval's start and start
 *     update the end to be the max of the current interval's end and end
 *   else:
 *     create a new interval containing start and end and add it to the merged list
 *     update the start to be the current interval's start
 *     update the end to be the current interval's end
 * add the start and end to the list
 * return the merged list
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(n)
 */
import java.util.*;

public class lc57 {
    public static int[][] insert(int[][] intervals, int[] newInterval) {
        if (intervals == null || intervals.length == 0) {
            return new int[][] { newInterval };
        }

        List<int[]> mergedIntervals = new ArrayList<>();

        int start = newInterval[0];
        int end = newInterval[1];

        for (int i = 0; i < intervals.length; i++) {
            int currentStart = intervals[i][0];
            int currentEnd = intervals[i][1];

            // In this case, the current interval is non-overlapping, add it to the list
            // Otherwise, (in the else if and else) follow the logic of LC 56 to merge
            // intervals and update start/end
            if (currentEnd < start) {
                mergedIntervals.add(new int[] { currentStart, currentEnd });
            } else if (currentStart <= end) {
                start = Math.min(start, currentStart);
                end = Math.max(end, currentEnd);
            } else {
                mergedIntervals.add(new int[] { start, end });
                start = currentStart;
                end = currentEnd;
            }
        }

        mergedIntervals.add(new int[] { start, end });

        return mergedIntervals.toArray(new int[mergedIntervals.size()][2]);
    }

    public static void main(String[] args) {
        int[][] input = new int[][] { { 1, 3 }, { 5, 7 }, { 8, 12 } };
        System.out.println("Intervals after inserting the new interval: "
                + Arrays.deepToString(insert(input, new int[] { 4, 6 })));

        System.out.println("Intervals after inserting the new interval: "
                + Arrays.deepToString(insert(input, new int[] { 4, 10 })));

        input = new int[][] { { 2, 3 }, { 5, 7 } };
        System.out.println("Intervals after inserting the new interval: "
                + Arrays.deepToString(insert(input, new int[] { 1, 4 })));
    }
}

File: joshuaCarpentier/assignments/mergeIntervals/lc57/lc57.js
// Problem Statement #

// Given a list of non-overlapping intervals sorted by their start time, insert a given interval at the correct position and merge 
// all necessary intervals to produce a list that has only mutually exclusive intervals.



class Interval {
  constructor(start, end) {
    this.start = start;
    this.end = end;
    // were using these in the rest of the code
  }

  print_interval() {
    process.stdout.write(`[${this.start}, ${this.end}]`);
  }
}


/**
 * Input: intervals Array and new inerval to insert
 * Output: return new merged interval
 * 
 * Intialize array and index;
 * Initialize while loop and set condition
 * 
 * condtion: while (index < intervals.length and Interval current position at the end is 
 * less then the new interval start)
 * 
 * push to the array we Initialized and increment index by 1 position 
 * 
 * Then Initialize next loop to merge all the intervals that overlap
 * 
 * condition: while (index < intervals.length and Interval current position at the start is 
 *   less then the new interval end)
 * 
 * we set the start of the interval and new interval to the minmum
 * and the end of the interval and new interval to the maximum
 * 
 * increment index by 1 position 
 * 
 * then push the new interval in the array we intiallized
 * 
 * and initalize a third condition
 * 
 * condition: which iterates through the intervals length and pushes the 
 * cureent interval index subarray in the array we intiallized and increment index by 1 position
 * 
 * and we return the array
 */
const insert = function (intervals, new_interval) {
  let merged = [],
      index = 0;
  
    // skip and add to output) all intervals that come before the 'new_interval'
    while (index < intervals.length && intervals[index].end < new_interval.start) {
      merged.push(intervals[index]);
      index++;
    }
  
    // merge all intervals that overlap with 'new_interval'
    while (index < intervals.length && intervals[index].start <= new_interval.end) {
      new_interval.start = Math.min(intervals[index].start, new_interval.start);
      new_interval.end = Math.max(intervals[index].end, new_interval.end);
      index++;
    }
  
    // insert the new_interval
    merged.push(new_interval);
  
    // add all the remaining intervals to the output
    while (index < intervals.length) {
      merged.push(intervals[index]);
      index++;
    }
  
    return merged;
};

// process.stdout.write('Intervals after inserting the new interval: ');
// let result = insert(
//   [],
//   new Interval(4, 6)
// );
// for (i = 0; i < result.length; i++) {
//   result[i].print_interval();
// }
// console.log();

process.stdout.write('Intervals after inserting the new interval: ');
let result = insert(
  [new Interval(1, 3), new Interval(5, 7), new Interval(8, 12)],
  new Interval(4, 6)
);
for (i = 0; i < result.length; i++) {
  result[i].print_interval();
}
console.log();

// process.stdout.write('Intervals after inserting the new interval: ');
// result = insert(
//   [new Interval(1, 3), new Interval(5, 7), new Interval(8, 12)],
//   new Interval(4, 10)
// );
// for (i = 0; i < result.length; i++) {
//   result[i].print_interval();
// }
// console.log();

// process.stdout.write('Intervals after inserting the new interval: ');
// result = insert([new Interval(2, 3), new Interval(5, 7)], new Interval(1, 4));
// for (i = 0; i < result.length; i++) {
//   result[i].print_interval();
// }
// console.log();

// Solution
// -----
// function insert(intervals, new_interval) {
//   let merged = [],
//     i = 0;

//   // skip and add to output) all intervals that come before the 'new_interval'
//   while (i < intervals.length && intervals[i].end < new_interval.start) {
//     merged.push(intervals[i]);
//     i += 1;
//   }

//   // merge all intervals that overlap with 'new_interval'
//   while (i < intervals.length && intervals[i].start <= new_interval.end) {
//     new_interval.start = Math.min(intervals[i].start, new_interval.start);
//     new_interval.end = Math.max(intervals[i].end, new_interval.end);
//     i += 1;
//   }

//   // insert the new_interval
//   merged.push(new_interval);

//   // add all the remaining intervals to the output
//   while (i < intervals.length) {
//     merged.push(intervals[i]);
//     i += 1;
//   }

//   return merged;
// }

// -----

// Time complexity #
// As we are iterating through all the intervals only once, the time complexity of the above algorithm is O(N), where ‘N’ is the total number of intervals.

// Space complexity #
// The space complexity of the above algorithm will be O(N) as we need to return a list containing all the merged intervals.


// const min = Math.min;
// const max = Math.max;

// const insert = function (intervals, new_interval) {
//   let inputLength = intervals.length;
//   let merged = [];
//   let startIdx; // store index value

//   // find merge start index
//   for (let i = 0; i < inputLength; i++) {
//     if(new_interval[0] <= intervals[i][1]){
//       startIdx = i;
//       break;
//     }
//   }

//   let left = intervals.slice(0, startIdx);
//   let endIdx = inputLength;
//   let [ll, rr] = [new_interval[0], new_interval[1]];

//   // merge process
//   for (let i = startIdx; i < inputLength; i++) {
//     let [el, er ] = [intervals[i][0], intervals[i][1]];
//     if(new_interval[1] < el){
//       // if target right < interval left merge ends
//       endIdx = i;
//       break;
//     }
//     ll = mi(ll, el); // each time merge left with min, right with max
//     rr = mx(rr, er);
//   }
//   let right = intervals.slice(endIdx);
//   merged = [...left, ...[[ll, rr]], ...right];

//   return merged;
// };